lang/py/src/avro/schemaNormalization.py

```diff
@@ -0,0 +1,211 @@
+        self.FP_TABLE=[]
+    
+        for i  in xrange (0,256):
+          fp = i;
+          for j in xrange(0,8):
+            mask = -(fp & 1L);
+            fp = (fp >> 1) ^ (self.EMPTY64 & mask);
+          self.FP_TABLE.append(fp)
+
+        self.env={}
+
+
+    def toParsingForm(self, schema):
+        """
+        Transform schema instance to its canonical json string .
+          @arg schema: avro schema instance
+          @return: the canonical form of the avro schema
+        """
+        self.env={}
+        return self._build(schema.to_json());
+
+    def fingerprint(self,fpName, data):
+        """
+        Process hash value of data using custom hash algorithm
+          @arg fpName: hash algorithm : CRC-64-AVRO or MD5 or SHA-1 or SHA-224 or SHA-256 or SHA-384 or SHA-512
+          @return: hash value in hexa
+        """
+        if (fpName.lower() == "crc-64-avro"):
+          fp = self.fingerprint64(data);
+          return fp
+    
+        m = None
+        if(fpName.lower() == "md5") :        m = hashlib.md5()
+        elif(fpName.lower() == "sha-1") :     m = hashlib.sha1()
+        elif(fpName.lower() == "sha-224") :   m = hashlib.sha224()
+        elif(fpName.lower() == "sha-256") :  m = hashlib.sha256()
+        elif(fpName.lower() == "sha-384") :  m = hashlib.sha384()
+        elif(fpName.lower() == "sha-512") :  m = hashlib.sha512()
+        
+        if m is not None:
+            m.update(data)
+            return m.hexdigest();
+        else:
+            raise NoSuchAlgorithmException("Unknown hash algorithm : '%s'"%(fpName))
+
+
+    def parsingFingerprint(self,fpName, schema):
+        """
+        Compute hash value of a schema canonical string using custom hash algorithm
+          @arg schema: avro schema instance
+          @return: the hash value of the canonical schema string
+        """
+        return self.fingerprint(fpName, self.toParsingForm(schema))
+
+    def parsingFingerprint64(self,schema):
+        """
+        Compute hash value of a schema canonical string using CRC-64-AVRO hash algorithm
+          @arg schema: avro schema instance
+          @return: the hash value of the canonical schema string
+        """
+        return self.fingerprint64(self.toParsingForm(schema))
+    
+    def fingerprint64(self,data):
+        """
+        Compute CRC-64-AVRO hash algorithm for a string
+          @arg data: string
+          @return: the hash value of the string
+        """
+        result = self.EMPTY64;
+        for b in data:
+          result = (result >> 8) ^ self.FP_TABLE[(result ^ ord(b)) & 0xff];
+        return hex(result)
+
+    def _build(self,jsonobj):
+        """
+        Create the canonical avro schema string (json) from a complete avro schema string
+        @arg json: json string representing an avro shema
+        @arg env: environment dict 
+        @return: canonical avro schema string
+        """
+        
+
+        #print jsonobj
```

Comment:
Please either remove commented out lines or uncomment and expressly wrap in a debug flag od some kind.